fw/http.c

```diff
@@ -3029,6 +3029,109 @@ tfw_h1_rewrite_purge_to_get(struct sk_buff *head)
 	return -ENOMEM;
 }
 
+#define CHOP_LEADING_CRLF
+#ifdef CHOP_LEADING_CRLF
+/**
+ * Removes single leading CR or CRLF or LF, if any.
+ * Indicates a buggy message if there are more leading CR or LF or other characters <= 0x20 with EBADMSG.
+ * @return zero on success and error code on error.
+ */
+static int
+tfw_h1_chop_leading_crlf(struct sk_buff **head)
+{
+	struct sk_buff *skb = *head;
+	struct sk_buff *stop = *head;
+	char *p, *q;
+	int i, n, ret, ba;
+	char buf[sizeof(__u32)];
+	struct skb_shared_info *si;
+	skb_frag_t *frag;
+
+	if (likely(skb_headlen(skb) >= 3))
+	{
+		p = q = skb->data;
+		if (unlikely(*p == \r)) p++;
+		if (unlikely(*p == \n)) p++;
+		if (unlikely(*p <= 0x20))
+			//Drop erroneous request
+			return EBADMSG;
+		if (unlikely((n = p - q)))
+		{
+			// Cleaning the msg
+			ret = ss_skb_chop_head_tail(skb, skb, n, 0);
+			if (ret)
+				return ret;
+		}
+		return 0;
+	}
+
+	// grab 3 bytes
+	n = 0; *(__u32*)&buf = 0;
+	do {
+		ba = skb_headlen(skb);
+		p = skb->data;
+		while (likely(ba--))
+		{
+			if (unlikely(n >= 3))
+				goto end_grab;
+			buf[n++] = p++;
+		}
+		
+		si = skb_shinfo(skb);
+		for (i = 0; i < si->nr_frags; i++)
+		{
+			frag = &si->frag[i];
+			ba = skb_frag_size(frag);
+			p = skb_frag_address(frag);
+			while (likely(ba--))
+			{
+				if (unlikely(n >= 3))
+					goto end_grab;
+				buf[n++] = p++;
+			}
+		}
+		skb = skb->next;
+	} while (skb != stop);
+	
+end_grab:
+	if (unlikely(n < 3)
+	{
+		// Impossible, however test for more correctness
+		return EBADMSG;
+	}
+	p = buf;
+	if (unlikely(*p == \r)) p++;
+	if (unlikely(*p == \n)) p++;
+	if (unlikely(*p <= 0x20))
+		//Drop erroneous request
+		return EBADMSG;
+	if (unlikely((n = p - buf)))
+	{
+		// Cleaning the msg
+		skb = *head;
+		while (n)
+		{
+			if (likely(skb->len > n)
+			{
+				return ss_skb_chop_head_tail(skb, skb, n, 0);
+			}
+			if (unlikely(skb->next == skb)
+			{
+				// the single skb in list
+				return EBADMSG;
+			}
+			n -= skb->len;
+			skb->next->prev = skn->prev;
+			skb->prev->next = skb->next;
+			*head = skb->next;
+			__kfree_skb(skb);
+			skb = *head;
+		}
+	}
```

Comment:
Well, I was wrong in our discussion when I proposed to move the cleaning logic to `tfw_h1_adjust_req()` state. What you do is actually _parse_ the [CR]LF, but you do this in very simple way. This parsing is way less efficient than our HTTP parser.

Now I see your point that the logic is completely different from `tfw_h1_rewrite_purge_to_get()` from #1533: in that case we just rewritten the data and we didn't need to parse anything. While in #1061 we can:
1. do the parsing as in the patch
2. set some flag that we need to remove the leading [CR]LF and also remember whether we need to cut off 1 or 2 bytes
3. cut off the bytes right in the parser

The current implementation is large, slow and buggy. We can clean it up, but there is no point to have the second parser. In ways (2) and (3) we just need to call `ss_skb_chop_head_tail()` with `n` = 1 or 2 according to the parser (remembered) state.

I still don't like the way (3) because we need to remove [CR]LF only for proxied requests, not for serviced from the case or blocked. Also the parser is one of the most performance crucial part and there is no reason to call the relatively heavyweight logic from it.

So let's add to flags (for 1 and 2 bytes cut off correspondingly) to the parse (see #1533) and just call `ss_skb_chop_head_tail()` in `tfw_h1_adjust_req()`, maybe with the wrapping function.

Reply:
@krizhanovsky 
1. Sorry, I missed the comment in time. Where are notifications? Now I know I should check gmail.

2. Thank you for remarks above. I replied to some and I will update the code for others.

3. I started to think on your new proposals. Probably I'll have some questions.
I have some minor questions & remarks just now

3.1. "just call ss_skb_chop_head_tail()" is not enough in case of request consisting of 1-byte chunks.
This is not always an attack. For instance, #1533 will block a request with 1-byte chunk in front
due to error code from ss_skb_chop_head_tail().

3.2. "do the parsing as in the patch" do you mean the current patch or the first one, which was inside http parser?

3.3. Where do we add flags?

Reply:
>  "just call ss_skb_chop_head_tail()" is not enough in case of request consisting of 1-byte chunks.

Yeah. Please add `WARN_ON(head || trail)` just before `return 0` in `ss_skb_chop_head_tail()` - we might have issues with extra small skbs in other places.

Then you just need to call function in a loop over the skb list. Probably you'll need to adjust the `ss_skb_chop_head_tail()` semantics, e.g. rename the current function to `__ss_skb_chop_head_tail()` and call it in a new `ss_skb_chop_head_tail()` in a loop over skb lists if the call didn't eat all the required bytes at the beginning or end of the list.

> Where do we add flags?

See TFW_HTTP_B_PURGE_GET in https://github.com/tempesta-tech/tempesta/pull/1533/files

> "do the parsing as in the patch" do you mean the current patch or the first one, which was inside http parser?

This current version with parsing in `tfw_h1_chop_leading_crlf()`

Reply:
Please double check that the proposed changes are the most optimal and consistent with the architecture.